# Derivative of cos(x) at x=1.0 radians

In [1]:
import math
import time

def dx(function,x,h):
    return (function(x+h)-function(x-h))/(2*h)

def threePointMidpointFormula(function, x):
    start_time = time.time()
    h = 0.000000001
    d = ((function(x+h)-function(x-h))/(2*h))
    end_time = time.time()
    total_time = end_time - start_time
    return d, total_time

def threePointEndpointFormula(function, x):
    start_time = time.time()
    h = 0.000000001
    d = ((-3*(function(x)))+(4*(function(x+h)))-(function(x+(2*h))))/((2*h))
    end_time = time.time()
    total_time = end_time - start_time
    return d, total_time    

def fivePointMidpointFormula(function, x):
    start_time = time.time()
    h = 0.000000001
    d = ((function(x-(2*h)))-(8*(function(x-h)))+(8*(function(x+h)))-(function(x+(2*h))))/((12*h))
    end_time = time.time()
    total_time = end_time - start_time
    return d, total_time

def fivePointEndpointFormula(function, x):
    start_time = time.time()
    h = 0.000000001
    d = ((-25*(function(x)))+(48*(function(x+h)))-(36*(function(x+(2*h))))+(16*(function(x+(3*h))))-(3*function(x+(4*h))))/((12*h))
    end_time = time.time()
    total_time = end_time - start_time
    return d, total_time

def richardsonDiff(function,x): 
    start_time = time.time()
    k=9
    L=[[0.0 for i in range(0,k)] for j in range(0,k)]
    for i in range(k):
        L[i][0]=dx(function,x,1/(2**(i+1)))
    for j in range(1,k):
        for i in range(k-j):
            L[i][j]=((4**(j))*L[i+1][j-1]-L[i][j-1])/(4**(j)-1)
    end_time = time.time()
    total_time = end_time - start_time
    return L[0][k-1], total_time


def main():
    f = lambda x: math.cos(x)
    x = 1.0

    exact=-math.sin(1.0)
    richard, richardTime = richardsonDiff(f,x)
    threePointMidpoint, threePointMidpointTime = threePointMidpointFormula(f,x)
    threePointEndpoint, threePointEndpointTime = threePointEndpointFormula(f,x)
    fivePointMidpoint, fivePointMidpointTime = fivePointMidpointFormula(f,x)
    fivePointEndpoint, fivePointEndpointTime = fivePointEndpointFormula(f,x)
    

    print(f'Exact Value:     \tValue: {exact}')
    print(f'Richardson:      \tValue: {richard}\tTime: {richardTime}')
    print(f'3-point Midpoint:\tValue: {threePointMidpoint}\tTime: {threePointMidpointTime}')
    print(f'3-point Endpoint:\tValue: {threePointEndpoint}\tTime: {threePointEndpointTime}')
    print(f'5-point Midpoint:\tValue: {fivePointMidpoint}\tTime: {fivePointMidpointTime}')
    print(f'5-point Endpoint:\tValue: {fivePointEndpoint}\tTime: {fivePointEndpointTime}')

    
main()

Exact Value:     	Value: -0.8414709848078965
Richardson:      	Value: -0.8414709848078831	Time: 0.00020956993103027344
3-point Midpoint:	Value: -0.8414710039872375	Time: 4.0531158447265625e-06
3-point Endpoint:	Value: -0.8414712260318424	Time: 4.0531158447265625e-06
5-point Midpoint:	Value: -0.8414709854835203	Time: 4.76837158203125e-06
5-point Endpoint:	Value: -0.8414713740615789	Time: 5.4836273193359375e-06


<style type="text/css">
.tg  {border-collapse:collapse;border-spacing:0;}
.tg td{font-family:Arial, sans-serif;font-size:14px;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:black;}
.tg th{font-family:Arial, sans-serif;font-size:14px;font-weight:normal;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:black;}
.tg .tg-c3ow{border-color:inherit;text-align:center;vertical-align:top}
</style>
<table class="tg">
  <tr>
    <th class="tg-c3ow"></th>
    <th class="tg-c3ow">3-point Endpoint</th>
    <th class="tg-c3ow">3-point Midpoint</th>
    <th class="tg-c3ow">5-point Endpoint</th>
    <th class="tg-c3ow">5-point Midpoint</th>
    <th class="tg-c3ow">Richard Extrapolation</th>
  </tr>
  <tr>
    <td class="tg-c3ow">Values</td>
    <td class="tg-c3ow">-0.8414712260318424</td>
    <td class="tg-c3ow">-0.8414710039872375</td>
    <td class="tg-c3ow">-0.8414713740615789</td>
    <td class="tg-c3ow">-0.8414709854835203</td>
    <td class="tg-c3ow">-0.8414709848078831</td>
  </tr>
  <tr>
    <td class="tg-c3ow">Exact Value</td>
    <td class="tg-c3ow">-0.8414709848078965</td>
    <td class="tg-c3ow">-0.8414709848078965</td>
    <td class="tg-c3ow">-0.8414709848078965</td>
    <td class="tg-c3ow">-0.8414709848078965</td>
    <td class="tg-c3ow">-0.8414709848078965</td>
  </tr>
  <tr>
    <td class="tg-c3ow">Times</td>
    <td class="tg-c3ow">4.0531158447265625e-06</td>
    <td class="tg-c3ow">3.0994415283203125e-06</td>
    <td class="tg-c3ow">5.0067901611328125e-06</td>
    <td class="tg-c3ow">4.291534423828125e-06</td>
    <td class="tg-c3ow">0.0001766681671142578</td>
  </tr>
  <tr>
    <td class="tg-c3ow">Absolute Error</td>
    <td class="tg-c3ow">2.4122394592929197e-07</td>
    <td class="tg-c3ow">1.9179341004260664e-08</td>
    <td class="tg-c3ow">3.892536823979498e-07</td>
    <td class="tg-c3ow">6.756237791449848e-10</td>
    <td class="tg-c3ow">-1.3433698597964394e-14</td>
  </tr>
  <tr>
    <td class="tg-c3ow">Relative Error</td>
    <td class="tg-c3ow">2.866692745595602e-05</td>
    <td class="tg-c3ow">2.279263446200881e-06</td>
    <td class="tg-c3ow">4.625869570810429e-05</td>
    <td class="tg-c3ow">8.02907991838557e-08</td>
    <td class="tg-c3ow">1.596454166631955e-12</td>
  </tr>
</table>

# Discussion/Analysis

Based on the five different methods used, there is a correlation between the time it takes to compute the result and the accuracy of the result. The methods with the lowest time (3-point Midpoint and Endpoint) also have the lowest accuracy and the method with the highest time (Richardson Extrapolation) has the highest accuracy. The difference between the methods is about two orders of magnitude. However, the runtimes of all the methods never exceed a thousandth of a second. The absolute error ranges about 4-7 orders of magnitude, with the Richardson Extrapolation method having the best accuracy. More notably in the 3 and 5 point methods, the midpoint methods increase in accuracy with the 5 point method having the highest accuracy of the point methods. Overall, the errors of the five methods align with the expected errors.